text/0000-multi-api-lifecycle-descriptor.md

```diff
@@ -0,0 +1,176 @@
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Platforms
+Platforms should use the contents of `lifecycle.toml` to set `io.buildpacks.lifecycle.version`, and `io.buildpacks.lifecycle.apis` labels when creating builders.
+
+The lifecycle will set `io.buildpacks.lifecycle.version` and `io.buildpacks.lifecycle.apis` on all future released lifecycle images.
+
+Platforms may use the contents of the `io.buildpacks.lifecycle.apis` label when deciding which `CNB_PLATFORM_API` to use, or validating whether a buildpack is compatible with a given builder image.
+
+### `pack`
+`pack inspect-builder` should display deprecated, supported, and experimental API information when it is available.
+
+## Backwards Compatibility
+
+Currently lifecycle API and verison information is contained in the `io.buildpacks.builder.metadata` label. The following is an abridged example label
+```
+{
+  "buildpacks": [],
+  "created-by": {}
+  "description": "",
+  "lifecycle": {
+    "version": "<lifecycle version>",
+    "api": {
+      "buildpack": "<buildpack API version>",
+      "platform": "<platform API version>"
+    }
+  },
+  "stack": {},
+}
+```
+For backwards compatibility `pack create-builder` should continue to populate the above label, setting `lifecycle.api.buildpack`, and `lifecycle.api.platform` to the corresponding supported `min`.
+
+This will allow platforms the read the metadata in it's existing format to update gracefully. Eventually this compatibility layer will be removed.
+
+The lifecycle descriptor file itself will not preserve backwards compatibility. Most platforms consume builders rather than directly consuming lifecycle artifacts, and therefore the builder label is the important compatibility concern. Those platforms that do consume lifecycle artifacts control which versions of the lifecycle they bring in and can update their behavior to parse the new format for newer lifecycles.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Understanding buildpack/platform API support will require more documentation and education.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- We could stick with a single `api.buildpack` and `api.platform` in the lifecycle descriptor and let platforms infer support ranges based on the guidelines outlined in [RFC-0041](https://github.com/buildpacks/rfcs/blob/main/text/0041-api-version-compat.md#non-breaking-0x-api-versions)
+- Instead of introducing deprecated APIs we could commit to never deprecating or removing support for older APIs
+
+# Prior Art
+[prior-art]: #prior-art
+
+* Experimental APIs are similar to [alpha level APIs](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning) in Kubernetes.
+* https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should we require platforms to turn on support for experimental APIs with an environment variable?
```

Comment:
Why would we need an environment variable for that?

Reply:
I was thinking that or have experimental versions not just be numbers, i.e. `1.0-alpha`

Reply:
I removed experimental APIs entirely. I think we should punt this discussion to a future RFC.